### ISSUE TITLE
Long-running tests: rationalize classification and enabling

### ISSUE THREAD

**Issue by bjacob** [iree-org/iree#14152]
At the moment, some long-running tests are tagged to opt out specifically of some slow testing configurations.

There are 3 issues with how we currently do this:
* It's not clear whether a test is marked `noasan` specifically because of an ASAN error or just because it's slow.
* We have to add tags for all slow configurations even if we just mean "it's slow" for each of them.
* We haven't rationalized the overall tradeoffs between test latency and coverage in a while (that I know).

So this morning I have rerun all tests, ignoring disabling-tags (meaning [here](https://sourcegraph.com/github.com/openxla/iree/-/blob/build_tools/cmake/iree_check_test.cmake?L9) I just make it return `FALSE`) , in two slow-configs: ASAN and TSAN (both times, enabling the sanitizer also in bytecode modules). Then ordering by descending times, here are the lists of slow tests:

* [Slow tests with ASAN](https://gist.github.com/bjacob/b2ab1dd3f0ab3e1f54f15140c408d9cf)
* [Slow tests with TSAN](https://gist.github.com/bjacob/3aeb3e95c2c226913a471f866779d2fe)

As discussed recently on a couple of PRs with @ScottTodd @GMNGeoffrey , we should use some standard Bazel `size` tag to flag large tests. I think we still want to rely on generic `tag`/`LABELS` as the container format, rather than creating a whole new `size` field that we need to propagate everywhere; we can convert from a `size=foo` tag to a proper `size` in Bazel specific places.

[`size`](https://bazel.build/reference/be/common-definitions#test.size) defaults to `medium` and allows two larger values: `large` and `enormous`. Looking at the above ASAN and TSAN logs, we can:

* Use `size=enormous` for very large tests that should be excluded from sanitizers, despite sanitizers not being our slowest config (emulators and non-opt are slower) and us really wanting to get sanitizer coverage. The clear outlier here that needs to be tagged `size=enormous` is `iree/tests/e2e/models/resnet50_fake_weights.mlir.test`.
* Use `size=large` for tests that take >= 3 seconds in TSAN or ASAN (usually TSAN is slower). We would currently still keep them enabled in sanitizers, and skip them in slower configs (emulators, non-opt).

All tests that are currently tagged `noriscv` or `optonly` or `hostonly` would be tagged `size=large`. 

Existing tags such as `notsan` `noasan` `noriscv` `optonly` `hostonly` would be generally dropped, but we would look into each of these to see which ones to retain, so after this, any such remaining tag would be really meaningful ("there is a specific reason why this test is host-only or has an ASAN issue").

As a side effect, a number of tests that have been unintentionally disabled in sanitizer configs just because they were slow in slower configs (emulator) would be reenabled there.

**Comment by GMNGeoffrey** [comment]
I'm looking at the individual PRs as well, but some high-level comments. Bazel did a poor job of ever defining what test size exactly means. The [docs](https://bazel.build/reference/be/common-definitions#test.size) refer to unit, integrations, and e2e tests, but the only things that is actually different about that them is the default timeout and planned RAM usage (I think not enforced. Just affects scheduling to avoid OOMs). At some point in the relatively recent past, some group of people at Google decided to redefine what test sizes mean to be more specific. Of course, this wasn't externalized, was overly tailored to people writing web apps, and was as far as I can tell mostly ignored by everyone actually writing tests. One of the things I like about `timeout` over `size` is that at least it means something specific: we're talking about the length of time the test takes to run. The precise mapping to timeouts is configurable with a default. I think looking at the default and the explanation of test sizes, we don't really have any tests that are "enormous".

I think it would make sense to instead use a symbolic timeout and then set the timeout per configuration. The current mapping to CMake timeouts just hardcodes the default values. I had envisioned instead using CMake variables and then setting those at configure time with the defaults varying along with things like turning on asan. We could even do something fancy like multiple config parameters contributing in a multiplicative way to alter the default timeouts (though that's probably not necessary). I still think this better conveys what we're trying to do when we skip tests also: we're not skipping the tests because they're "big"; we're skipping them because they are expected to take up to X time (where X could just be "long").

What would change about your plan if we called the resnet test "long", the "slow on sanitizers" tests "moderate" and most of our tests "short". Is the issue that "medium" is the default?

I have an ulterior motive for not wanting to label things `enormous` as well, which I'd like to call out: Google's internal low-config CI doesn't run enormous tests. They're generally expected to be the sort of thing that is spinning up a server or something similarly heavyweight. So it would be slightly annoying for me to have to deal with us labeling things "enormous" externally as I'll need to remap it internally. That is a secondary concern though.

**Comment by GMNGeoffrey** [comment]
Urgh and of course I used the size terminology instead of the timeout terminology above. Edited to fix

**Comment by bjacob** [comment]
>  Is the issue that "medium" is the default?

Yes, I structured this around the default being `medium`, trying to keep the vast majority of tests on the default so that we wouldn't have to start providing a value for all tests.

But since we have our own rules, we can have them default to `small`.  So that wasn't a strong reason to stick to `medium`. I just.... defaulted to keeping our defaults aligned with the defaults of standard Bazel rules, but now I understand we have a reason to deviate.

Point taken that `timeout` usefully conveys that it's about time as opposed to other metrics such as RAM usage.  But the `out` part of `timeout` is IMHO unwanted.  It suggests that the primary reason why we set a particular `timeout` value on a test is to avoid the test... timing out.  But if we have a high-latency test, that's a productivity drain and a waste of resources no matter what the test outcome is. So here, we care about `time` not as a matter of avoiding `timeout`'s but as a measure of a cost. In that sense, `size` seems closer to the mark than `timeout` --- it does not convey the specific quantity we are trying to be economical with (time) but at least it does convey we are trying to be economical with some quantity, as opposed to `timeout` which conveys not caring about that, only caring about ultimate test status.

See how my PR does away altogether with the bazel-to-cmake mapping of timeout values in seconds. And we're not missing that, because we never wanted to get close to these timeouts in the first place. Everyone hates that 60-second test that is still running in CTest long after everything else is finished, making a mockery of our 128-core cloudtops now forced to wait on a single thread.

Summary - how about: 
 - stick to `size` not `timeout`
 - change default value of `size` to `small` in all our own Bazel and CMake rules
 - shift my original PR's use of `large` and `enormous` to `medium` and `large` instead.

**Comment by GMNGeoffrey** [comment]
The major benefit we got from adding timeouts to cmake is that a test becoming ridiculously slow was a fast-ish failure rather than until the entire job timed out or something. I added a backstop default to the cmake invocations on CI, so at least that stopped happening, but it has to accommodate the slowest test, so isn't very tight. So I would like us to continue setting per-test timeouts in cmake as well, even if they're mostly the default. Given that, it doesn't seem like having the *separate* concept of size is that helpful. I see your point that the timeout parameter sort of implies that we're "giving" the test that long, but I think it's pretty coherent to say that this test has a "long" timeout and also that we don't run tests with a "long" timeout with sanitizers.

I guess one advantage to a separate property would be if we decide that some test has a long timeout but it's still worth running with sanitizers. Then it would be helpful to be able to mark it as "medium" anyway. But then I think it's really not about size either: it's about cost-benefit tradeoffs. I think we probably could cross that bridge when we come to it. We could even have a separate invocation that explicitly lists these super special slow but valuable tests. But probably we should just make the test faster 🙂

**Comment by bjacob** [comment]
ok, i will revert to `timeout`.
